File: src/1039_Minimum_Score_Triangulation_of_Polygon.py
# 1039_Minimum_Score_Triangulation_of_Polygon
# https://leetcode.cn/problems/minimum-score-triangulation-of-polygon/submissions/

class Solution:
    def minScoreTriangulation(self, values: List[int]) -> int:
        n = len(values)
        dp = [[0] * n for _ in range(n)]

        # 初始化 dp 数组
        for i in range(n - 1):
            dp[i][i + 1] = 0

        # 逐步增加多边形的大小
        for l in range(2, n):
            for i in range(n - l):
                j = i + l
                dp[i][j] = float('inf')
                # 遍历中间顶点 k
                for k in range(i + 1, j):
                    dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j] + values[i] * values[k] * values[j])

        return dp[0][n - 1]


class Solution:
    def minScoreTriangulation(self, values: List[int]) -> int:
        @cache
        def dfs(i: int, j: int) -> int:
            if i + 1 == j:
                return 0
            return min(dfs(i, k) + dfs(k, j) + values[i] * values[k] * values[j] for k in range(i + 1, j))

        return dfs(0, len(values) - 1)
    

class Solution:
    def minScoreTriangulation(self, v: List[int]) -> int:
        n = len(v)
        f = [[0] * n for _ in range(n)]
        for i in range(n - 3, -1, -1):
            for j in range(i + 2, n):
                f[i][j] = min(f[i][k] + f[k][j] + v[i] * v[j] * v[k]
                              for k in range(i + 1, j))
        return f[0][-1]

File: src/1039_Minimum_Score_Triangulation_of_Polygon.rs
// 1039_Minimum_Score_Triangulation_of_Polygon
// https://leetcode.cn/problems/minimum-score-triangulation-of-polygon/

use std::collections::HashMap;
impl Solution {
    pub fn min_score_triangulation(values: Vec<i32>) -> i32 {
        fn dfs(i: usize, j: usize, values: &[i32], memo: &mut HashMap<(usize, usize), i32>) -> i32 {
            if i + 1 == j {
                return 0;
            }
            if let Some(&result) = memo.get(&(i, j)) {
                return result;
            }
            let result = (i + 1..j)
                .map(|k| dfs(i, k, values, memo) + dfs(k, j, values, memo) + values[i] * values[k] * values[j])
                .min()
                .unwrap();
            memo.insert((i, j), result);
            result
        }

        let mut memo = HashMap::new();
        dfs(0, values.len() - 1, &values, &mut memo)
    }
}


impl Solution {
    pub fn min_score_triangulation(values: Vec<i32>) -> i32 {
        fn dfs(i: usize, j: usize, v: &[i32], tab: &mut Vec<Vec<i32>>) -> i32 {
            if tab[i][j] != i32::MAX {
                return tab[i][j];
            }
            for k in i + 1..j {
                tab[i][j] = tab[i][j].min(dfs(i, k, v, tab) + dfs(k, j, v, tab) + v[i] * v[k] * v[j]);
            }
            tab[i][j]
        }

        let n = values.len();
        let mut tab = vec![vec![i32::MAX; n]; n];

        for i in 0..n - 1 {
            tab[i][i + 1] = 0; // 两个点无法组成三角形
        }

        dfs(0, n - 1, &values, &mut tab)
    }
}
